drivers/watchdog/wdt_wwdg_stm32.c

```diff
@@ -0,0 +1,241 @@
+ *             = 1048.576 ms
+ */
+
+#define ABS_DIFF_UINT(a, b)  ((a) > (b) ? (a) - (b) : (b) - (a))
+#define WWDG_TIMEOUT_ERROR_MARGIN   (100 * USEC_PER_MSEC)
+#define IS_WWDG_TIMEOUT(__TIMEOUT_GOLDEN__, __TIMEOUT__)  \
+	(ABS_DIFF_UINT(__TIMEOUT_GOLDEN__, __TIMEOUT__) < \
+	 WWDG_TIMEOUT_ERROR_MARGIN)
+
+static void wwdg_stm32_irq_config(struct device *dev);
+
+/**
+ * @brief Calculates the timeout in microseconds.
+ *
+ * @param prescaler The prescaler value.
+ * @param counter The counter value.
+ * @return The timeout calculated in microseconds.
+ */
+static u32_t wwdg_stm32_get_timeout(u32_t prescaler, u32_t counter)
+{
+	u32_t divider = WWDG_INTERNAL_DIVIDER * (1 << (prescaler >> 7));
+	float f_wwdg = SystemCoreClock / divider;
+
+	return USEC_PER_SEC * (((counter & 0x3F) + 1) / f_wwdg);
+}
+
+/**
+ * @brief Calculates prescaler & counter values.
+ *
+ * @param timeout Timeout value in microseconds.
+ * @param prescaler Pointer to prescaler value.
+ * @param counter Pointer to counter value.
+ */
+static void wwdg_stm32_convert_timeout(u32_t timeout,
+				       u32_t *prescaler,
+				       u32_t *counter)
+{
+	u32_t clock_freq = SystemCoreClock;
+	u8_t divider = 0U;
+	u8_t shift = 3U;
+
+	/* Convert timeout to seconds. */
+	float timeout_s = (float)timeout / USEC_PER_SEC;
+	float wwdg_freq;
+
+	*prescaler = 0;
+	*counter = 0;
+
+	for (divider = 8; divider >= 1; divider >>= 1) {
+		wwdg_freq = ((float)clock_freq) / WWDG_INTERNAL_DIVIDER / divider;
+		/* +1 to ceil the result, which may lose from truncation */
+		*counter = (u32_t)(timeout_s * wwdg_freq + 1) - 1;
+		*counter |= WWDG_RESET_LIMIT;
+		*prescaler = shift << 7;
+
+		if (*counter <= WWDG_COUNTER_MAX) {
+			break;
+		}
+
+		shift--;
+	}
+}
+
+static int wwdg_stm32_setup(struct device *dev, u8_t options)
+{
+	WWDG_TypeDef *wwdg = WWDG_STM32_STRUCT(dev);
+
+	/* Deactivate running when debugger is attached. */
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+#if defined(CONFIG_SOC_SERIES_STM32F0X)
+		LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_DBGMCU);
+#elif defined(CONFIG_SOC_SERIES_STM32L0X)
+		LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_DBGMCU);
+#endif
+		LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_WWDG_STOP);
+	}
+
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		return -ENOTSUP;
+	}
+
+	/* Enable the WWDG */
+	LL_WWDG_Enable(wwdg);
+
+	return 0;
+}
+
+static int wwdg_stm32_disable(struct device *dev)
+{
+	/* watchdog cannot be stopped once started unless SOC gets a reset */
+	ARG_UNUSED(dev);
+
+	return -EPERM;
+}
+
+static int wwdg_stm32_install_timeout(struct device *dev,
+				      const struct wdt_timeout_cfg *config)
+{
+	struct wwdg_stm32_data *data = WWDG_STM32_DATA(dev);
+	WWDG_TypeDef *wwdg = WWDG_STM32_STRUCT(dev);
+	u32_t timeout = config->window.max * USEC_PER_MSEC;
+	u32_t calculated_timeout;
+	u32_t prescaler = 0U;
+	u32_t counter = 0U;
+
+	if (config->callback != NULL) {
+		data->callback = config->callback;
+		wwdg_stm32_irq_config(dev);
```

Comment:
We generally call IRQ init function in device init function.
Calling it here, you'll call multiple times `IRQ_CONNECT` and `irq_enable`. 
I don't think this should be harmful, but this is not usual.